File: examples/reference-implementations/python-client/ossa_client/agents.py
"""
OSSA Agent Operations

Client methods for agent discovery, publishing, and lifecycle management.
"""

from typing import Any, Dict, List, Optional
from .client import OSSAClient


class AgentClient:
    """
    Agent operations client for OSSA registry.

    Provides methods for searching, publishing, and managing agents.
    """

    def __init__(self, client: OSSAClient):
        self._client = client

    def search(
        self,
        q: Optional[str] = None,
        tag: Optional[str] = None,
        capability: Optional[str] = None,
        domain: Optional[str] = None,
        publisher: Optional[str] = None,
        license: Optional[str] = None,
        compliance: Optional[str] = None,
        verified: Optional[bool] = None,
        min_rating: Optional[float] = None,
        sort: str = 'relevance',
        limit: int = 20,
        offset: int = 0,
    ) -> Dict[str, Any]:
        """
        Search and list agents.

        Args:
            q: Full-text search query
            tag: Filter by tag
            capability: Filter by capability
            domain: Filter by domain
            publisher: Filter by publisher
            license: Filter by license (SPDX identifier)
            compliance: Filter by compliance profile
            verified: Only verified publishers
            min_rating: Minimum rating (1-5)
            sort: Sort order (downloads, rating, updated, created, relevance)
            limit: Results per page (max 100)
            offset: Pagination offset

        Returns:
            Search results with agents list and metadata
        """
        return self._client.request(
            'GET',
            '/agents',
            query={
                'q': q,
                'tag': tag,
                'capability': capability,
                'domain': domain,
                'publisher': publisher,
                'license': license,
                'compliance': compliance,
                'verified': verified,
                'min_rating': min_rating,
                'sort': sort,
                'limit': limit,
                'offset': offset,
            },
        )

    def publish(
        self,
        manifest: Dict[str, Any],
        package: Dict[str, Any],
        license: str,
        documentation: Optional[Dict[str, str]] = None,
        keywords: Optional[List[str]] = None,
        dependencies: Optional[Dict[str, str]] = None,
    ) -> Dict[str, Any]:
        """
        Publish a new agent or version.

        Args:
            manifest: OSSA agent manifest
            package: Package information (tarball_url, shasum, size_bytes)
            license: SPDX license identifier
            documentation: Documentation URLs (readme, changelog, repository)
            keywords: Agent keywords (max 20)
            dependencies: Agent dependencies

        Returns:
            Publish response with status and verification info
        """
        return self._client.request(
            'POST',
            '/agents',
            body={
                'manifest': manifest,
                'package': package,
                'license': license,
                'documentation': documentation,
                'keywords': keywords,
                'dependencies': dependencies,
            },
        )

    def get(self, publisher: str, name: str) -> Dict[str, Any]:
        """
        Get agent details (latest version).

        Args:
            publisher: Publisher identifier
            name: Agent name

        Returns:
            Agent details including metadata, stats, and versions
        """
        return self._client.request('GET', f'/agents/{publisher}/{name}')

    def get_version(self, publisher: str, name: str, version: str) -> Dict[str, Any]:
        """
        Get specific agent version.

        Args:
            publisher: Publisher identifier
            name: Agent name
            version: Semantic version

        Returns:
            Agent version details
        """
        return self._client.request('GET', f'/agents/{publisher}/{name}/{version}')

    def list_versions(self, publisher: str, name: str) -> Dict[str, Any]:
        """
        List all versions of an agent.

        Args:
            publisher: Publisher identifier
            name: Agent name

        Returns:
            List of agent versions with metadata
        """
        return self._client.request('GET', f'/agents/{publisher}/{name}/versions')

    def unpublish(
        self,
        publisher: str,
        name: str,
        version: str,
        reason: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Unpublish a specific agent version.

        Args:
            publisher: Publisher identifier
            name: Agent name
            version: Semantic version
            reason: Reason for unpublishing

        Returns:
            Unpublish confirmation
        """
        body = {'reason': reason} if reason else None
        return self._client.request('DELETE', f'/agents/{publisher}/{name}/{version}', body=body)

    def deprecate(
        self,
        publisher: str,
        name: str,
        version: str,
        reason: str,
        replacement_version: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Deprecate a specific agent version.

        Args:
            publisher: Publisher identifier
            name: Agent name
            version: Semantic version to deprecate
            reason: Reason for deprecation
            replacement_version: Recommended replacement version

        Returns:
            Deprecation confirmation
        """
        return self._client.request(
            'POST',
            f'/agents/{publisher}/{name}/{version}/deprecate',
            body={
                'reason': reason,
                'replacement_version': replacement_version,
            },
        )

    def get_dependencies(
        self,
        publisher: str,
        name: str,
        version: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Get agent dependencies.

        Args:
            publisher: Publisher identifier
            name: Agent name
            version: Specific version (default: latest)

        Returns:
            Dependencies with version resolution and dependency tree
        """
        query = {'version': version} if version else None
        return self._client.request('GET', f'/agents/{publisher}/{name}/dependencies', query=query)

    def get_stats(
        self,
        publisher: str,
        name: str,
        period: str = 'all',
    ) -> Dict[str, Any]:
        """
        Get agent download and usage statistics.

        Args:
            publisher: Publisher identifier
            name: Agent name
            period: Time period (day, week, month, year, all)

        Returns:
            Statistics including downloads and installs
        """
        return self._client.request('GET', f'/agents/{publisher}/{name}/stats', query={'period': period})

File: examples/reference-implementations/typescript-client/src/agents.ts
/**
 * OSSA Agent Operations
 *
 * Client methods for agent discovery, publishing, and lifecycle management.
 */

import { OSSAClient } from './client.js';

export interface AgentManifest {
  apiVersion: string;
  kind: 'Agent';
  metadata: {
    name: string;
    version: string;
    description: string;
    labels?: Record<string, string>;
  };
  spec: {
    taxonomy: {
      domain: string;
      subdomain?: string;
      capability?: string;
    };
    role: string;
    llm?: {
      provider: string;
      model: string;
      temperature?: number;
      maxTokens?: number;
    };
    capabilities?: Array<{
      name: string;
      description: string;
      input_schema: Record<string, unknown>;
      output_schema: Record<string, unknown>;
    }>;
    runtime?: {
      type: string;
      config?: Record<string, unknown>;
    };
  };
}

export interface PublishRequest {
  manifest: AgentManifest;
  package: {
    tarball_url: string;
    shasum: string;
    size_bytes: number;
  };
  documentation?: {
    readme?: string;
    changelog?: string;
    repository?: string;
  };
  license: string;
  keywords?: string[];
  dependencies?: Record<string, string>;
}

export interface PublishResponse {
  status: 'published' | 'pending' | 'failed';
  agent: {
    name: string;
    version: string;
    publisher: string;
    published_at: string;
    registry_url: string;
    package_url: string;
  };
  verification: {
    schema_valid: boolean;
    security_scan: 'passed' | 'failed' | 'pending';
    verified_publisher: boolean;
  };
}

export interface SearchParams {
  q?: string;
  tag?: string;
  capability?: string;
  domain?: string;
  publisher?: string;
  license?: string;
  compliance?: string;
  verified?: boolean;
  min_rating?: number;
  sort?: 'downloads' | 'rating' | 'updated' | 'created' | 'relevance';
  limit?: number;
  offset?: number;
}

export interface AgentSearchResult {
  name: string;
  version: string;
  publisher: string;
  description: string;
  license: string;
  downloads: number;
  rating: number;
  verified: boolean;
  tags: string[];
  capabilities: string[];
  created_at: string;
  updated_at: string;
  registry_url: string;
}

export interface SearchResponse {
  total: number;
  limit: number;
  offset: number;
  agents: AgentSearchResult[];
}

export interface AgentDetails extends AgentSearchResult {
  long_description: string;
  repository: string;
  homepage: string;
  documentation: string;
  taxonomy: {
    domain: string;
    subdomain?: string;
    capability?: string;
  };
  compliance_profiles: string[];
  rating_info: {
    average: number;
    count: number;
  };
  download_stats: {
    total: number;
    last_month: number;
    last_week: number;
  };
  versions: Array<{
    version: string;
    published_at: string;
    changelog?: string;
  }>;
  dependencies: Record<string, string>;
  manifest_url: string;
  package_url: string;
}

export interface VersionInfo {
  version: string;
  published_at: string;
  downloads: number;
  changelog_url?: string;
  manifest_url: string;
  package_url: string;
  deprecated?: boolean;
  deprecation_reason?: string;
}

export interface VersionsResponse {
  agent: string;
  versions: VersionInfo[];
}

/**
 * Agent operations client
 */
export class AgentClient {
  constructor(private client: OSSAClient) {}

  /**
   * Search and list agents
   */
  async search(params: SearchParams = {}): Promise<SearchResponse> {
    return this.client.request<SearchResponse>({
      method: 'GET',
      path: '/agents',
      query: params as Record<string, string | number | boolean>,
    });
  }

  /**
   * Publish a new agent or version
   */
  async publish(request: PublishRequest): Promise<PublishResponse> {
    return this.client.request<PublishResponse>({
      method: 'POST',
      path: '/agents',
      body: request,
    });
  }

  /**
   * Get agent details (latest version)
   */
  async get(publisher: string, name: string): Promise<AgentDetails> {
    return this.client.request<AgentDetails>({
      method: 'GET',
      path: `/agents/${publisher}/${name}`,
    });
  }

  /**
   * Get specific agent version
   */
  async getVersion(publisher: string, name: string, version: string): Promise<AgentDetails> {
    return this.client.request<AgentDetails>({
      method: 'GET',
      path: `/agents/${publisher}/${name}/${version}`,
    });
  }

  /**
   * List all versions of an agent
   */
  async listVersions(publisher: string, name: string): Promise<VersionsResponse> {
    return this.client.request<VersionsResponse>({
      method: 'GET',
      path: `/agents/${publisher}/${name}/versions`,
    });
  }

  /**
   * Unpublish a specific agent version
   */
  async unpublish(
    publisher: string,
    name: string,
    version: string,
    reason?: string
  ): Promise<{ status: string; agent: string; version: string; unpublished_at: string }> {
    return this.client.request({
      method: 'DELETE',
      path: `/agents/${publisher}/${name}/${version}`,
      body: reason ? { reason } : undefined,
    });
  }

  /**
   * Deprecate a specific agent version
   */
  async deprecate(
    publisher: string,
    name: string,
    version: string,
    reason: string,
    replacementVersion?: string
  ): Promise<{
    status: string;
    agent: string;
    version: string;
    deprecated_at: string;
    replacement_version?: string;
  }> {
    return this.client.request({
      method: 'POST',
      path: `/agents/${publisher}/${name}/${version}/deprecate`,
      body: {
        reason,
        replacement_version: replacementVersion,
      },
    });
  }

  /**
   * Get agent dependencies
   */
  async getDependencies(
    publisher: string,
    name: string,
    version?: string
  ): Promise<{
    agent: string;
    version: string;
    dependencies: {
      runtime?: Record<string, unknown>;
      agents?: Record<string, unknown>;
      tools?: Record<string, unknown>;
    };
    dependency_tree: Record<string, unknown>;
  }> {
    return this.client.request({
      method: 'GET',
      path: `/agents/${publisher}/${name}/dependencies`,
      query: version ? { version } : undefined,
    });
  }
}
